### PR TITLE
auto_init netif: conditionally initialize GNRC

### PR DESCRIPTION
**Pull request by OlegHahm**
Enable auto initialization of network interfaces with or without GNRC.

~~Depends on #5185.~~

### CODE REVIEW

**Comment by OlegHahm** [comment]
(Sorry, just set the label to check the CI results if I forget anything.)

**Comment by authmillenon** [comment]
How is one supposed to get the pointers to the netdev?

**Comment by authmillenon** [comment]
In general: thanks for the effort to find a solution for #4672, but I think it's not as simple :/:

* the pointer problem (which could be fixed easily)
* the MAC layer problem: how to define which MAC implementation to use with GNRC.

**Comment by OlegHahm** [comment]
> the pointer problem (which could be fixed easily)

Will do.

> the MAC layer problem: how to define which MAC implementation to use with GNRC.

Well, unless we have any MAC implementation we don't have to deal with it, but in my working branch for TSCH I simply use the same mechanism we use for whole auto initialization currently: ifdefs.

**Comment by authmillenon** [comment]
> Well, unless we have any MAC implementation we don't have to deal with it, but in my working branch for TSCH I simply use the same mechanism we use for whole auto initialization currently: ifdefs.

So interfaces with different MAC protocols are not possible?

**Comment by kaspar030** [comment]
> So interfaces with different MAC protocols are not possible?

Like, at the same time? Let's also not forget about IPv7!

**Comment by OlegHahm** [comment]
> So interfaces with different MAC protocols are not possible?

Not with auto-initialization. But we do not even have a single MAC protocol yet, so I think we can re-design once we're there and go with the simple solution for now.

**Comment by OlegHahm** [comment]
Updated. Now exports a variable called `netdev_default` and renamed `auto_init_gnrc_netif` into `auto_init_netdev`.

**Comment by OlegHahm** [comment]
Oops, just realized that `netdev_default` just be independent from `auto_init_netdev`. Will fix.

**Comment by authmillenon** [comment]
>>  So interfaces with different MAC protocols are not possible?
>
> Like, at the same time? Let's also not forget about IPv7!

I don't understand why you frame it as something unrealistic/undesirable... Given that there are a number of MAC protocols out there for WSN (e.g. ContikiMAC, IEEE 802.15.4e, ... just look at the proposed MACs in the PRs), I don't think it is a bad or undesirable to have a RIOT node border route between two different MAC'd networks of the same PHY layer...

**Comment by authmillenon** [comment]
> Updated. Now exports a variable called netdev_default and renamed auto_init_gnrc_netif into auto_init_netdev.

Who determines what netdev will become `netdev_default`? How do I get other netdevs when I need them? Wouldn't it be better to export an array with pointers to ALL netdevs?

**Comment by authmillenon** [comment]
Though what order do they come in then?

**Comment by OlegHahm** [comment]
> Who determines what netdev will become netdev_default? How do I get other netdevs when I need them? Wouldn't it be better to export an array with pointers to ALL netdevs? 

`netdev_default` is the same as before. A pointer to all the devices could be provided, too - but might go into a different PR.

**Comment by authmillenon** [comment]
> netdev_default is the same as before.

"same as before" compared to what? The module? How do you get this information from the build system? What instance of the module are you choosing if multiple devices of that type are available?

Moreover, if I need to add auto_init to the tests for #3551 and #4713 I at least don't need them exposed I realized now. Since initialization for the stacks should happen there as well.

**Comment by OlegHahm** [comment]
> "same as before" compared to what? The module? How do you get this information from the build system?

Same as before this PR. The Makefile in the board directory determines that.

> What instance of the module are you choosing if multiple devices of that type are available?

The first one, determined by the params configuration.

**Comment by OlegHahm** [comment]
> Oops, just realized that netdev_default just be independent from auto_init_netdev. Will fix.

#5183 is required for that.

**Comment by OlegHahm** [comment]
Okay, how about this: let the board define the default parameter set (identifying the default transceiver in case several configurations exist) and expose a `netdev_default_setup()` function in case one uses `netdev_default` without `auto_init`.

File: boards/native/Makefile.dep [hunk 1]
```diff
@@ -1,3 +1,4 @@
 ifneq (,$(filter netdev_default gnrc_netdev_default,$(USEMODULE)))
     USEMODULE += netdev2_tap
+	CFLAGS += -DNETDEV_DEFAULT_TAP
```
Comment (authmillenon) [inline boards/native/Makefile.dep]: indentation.

**Comment by OlegHahm** [comment]
Rebased to current master - should be ready to be merged.

**Comment by kaspar030** [comment]
@OlegHahm I know I'm late, but one global "netdev_default"? Could we give this change some time to discuss?

**Comment by OlegHahm** [comment]
@kaspar030, sure, we can give it time, but what's the issue?

**Comment by kaspar030** [comment]
Plus hundred twenty lines, a lot of defines, same global variable put in scope depending on defines. At first look it doesn't scale, seems overly complicated.

This seems to me like a shortcut around everything discussed regarding generic auto-initalization.

It is probably ingenius, so I'd like to have more time to take a closer look!

**Comment by OlegHahm** [comment]
> It is probably ingenius, so I'd like to have more time to take a closer look!

It's probably neither ingenious nor completely stupid - it's just an addition to the current auto_init that allows to initialize the default network device of a board without GNRC and provides a pointer to it (plus a generic function to set it up).

> This seems to me like a shortcut around everything discussed regarding generic auto-initalization.

I agree - it's more like a workaround unless we have a fancy solution. But at least it leaves linkerscripts untouched. ;)

**Comment by OlegHahm** [comment]
Should have read more carefully:

@authmillenon wrote:
> In general: thanks for the effort to find a solution for #4672, but I think it's not as simple :/:

I thought you were talking about #4713.

Anyway, I agree that this might be too much for today and the test application in #4713 may stay a little bit hacky for this release.

**Comment by OlegHahm** [comment]
Btw. I might close or completely overhaul this PR - I have a (slightly) better idea in mind.

**Comment by miri64** [comment]
Did you?

**Comment by kYc0o** [comment]
Will postpone since @OlegHahm seems to have a nicer idea :P

**Comment by miri64** [comment]
Can be closed as memo?

**Comment by OlegHahm** [comment]
Can be closed without a memo, I completely forgot my solution.

**Pull request closed without merging**